### DISCUSSION TITLE
Initial pose setting problem

### DISCUSSION THREAD

**Discussion by Lilly-yang** [prl-mushr/mushr#28] [Software]
Hi,

I am currently studying [Navigation Tutorial](https://mushr.io/tutorials/navigation/). I failed to set the initial posture of the car on the map. It will always change automatically. (Here are the [screen vedio](https://drive.google.com/file/d/1s9vDMvegCe7DPRmr2GHH5uUtRLqu4Gso/view?usp=sharing) and real scene)
![IMG_9513](https://user-images.githubusercontent.com/39558023/120342194-d9b28c80-c2f7-11eb-8d62-dfd0ceabdd6e.jpg)

And (maybe because of the initial pose problem) I also can't set the goal pose.

Could you give me some suggestions on why it happened and how to fix it? Thank you!

**Comment by schmittlema** [comment]
So after taking a look at the video it looks like your map doesn't line up with the room you are operating in. The navigation tutorial relies on  [mushr_pf](https://github.com/prl-mushr/mushr_pf) which is a particle filter. The particle filter uses the laser scanner to localize (moving dots in your video). Basically it tries to match the current laser scan with the map of the environment. And since your map is of the track and not the room, the particle filter does not localize. To build a map of the room try out our new [tutorial on map building ](https://mushr.io/tutorials/mapping/). 

I'm assuming you are using the track map so that the rhc controller stays within the path. That makes a lot of sense, but you will need to have a separate map for localization (following the mapping tutorial) and for control (your current map) in order for it to work. So you will want to publish both maps with separate [map_servers](http://wiki.ros.org/map_server). This can be done by running `rosrun map_server map_server PATH_TO_YOUR_MAP.yaml` for each map. You will also need to have separate map topics so that your two map servers don't both publish to `/map`. See [here](https://answers.ros.org/question/227776/how-to-change-the-topic-name-for-publishing-occupancy-grid-using-map_server/) for how to do that. And finally, mushr_rhc and mushr_pf need to be told to subscribe to their associated map. You can change the map topic for mushr_pf [here](https://github.com/prl-mushr/mushr_pf/blob/master/src/mushr_pf/particle_filter.py#L22), and you can leave mushr_rhc's map_topic to the default `/map`. Appologies that these topic names aren't parameters, that is on the todo list.

**Reply by Lilly-yang** [reply]
Thanks Matt.

I ran into a problem while trying to build a map of our room. After I ran [slam.launch](https://github.com/prl-mushr/mushr_base/blob/master/mushr_base/launch/includes/slam.launch), it seemed that the `tf tree` was disturbed. The topics on `rviz` cannot be transformed correctly.  (Refer to [recording](https://drive.google.com/file/d/1OFdlIiFLbjs_tMoYSOw_VenxyComkLZ0/view?usp=sharing)). This resulted in the loss of appearance of the robot racing car and the LiDAR laser point. And there is no map generated when I run the car. I’m not sure if the ‘car/map’ is displayed correctly, it looks a little strange.

The following are trees after only running [teleop.launch](https://github.com/prl-mushr/mushr_base/blob/master/mushr_base/launch/teleop.launch) and after running  [slam.launch](https://github.com/prl-mushr/mushr_base/blob/master/mushr_base/launch/includes/slam.launch). You can see that a `car/map` has replaced the `map`, and the `map` is no longer connected to anything.
[full image](https://drive.google.com/file/d/1m4mQWOz64hJE2tZsuk75YyCcAw2NayYD/view?usp=sharing)
![image](https://user-images.githubusercontent.com/39558023/120682437-be34b680-c49c-11eb-9420-88e9a4fd6805.png)
[full image](https://drive.google.com/file/d/1KvRTHvfdQw81BnZlJzlXCZ9tnVKNltdj/view?usp=sharing)
![image](https://user-images.githubusercontent.com/39558023/120682530-d9072b00-c49c-11eb-995c-f876945f9860.png)

PS: All the `MuSHR` files I used have been updated.

I searched for some solutions on the Internet, but they all seem a bit complicated and I am not sure if it fits my problem. Sorry, I am not that familiar with `ROS` yet. Could you please give me some suggestions? Thank you!

**Reply by MarkusSchiffer** [reply]
Hey Lilly,

In RViz, can you try changing the Global Options/Fixed Frame from `/map` to `/car/map`? Hopefully that resolves your issue!

**Reply by Lilly-yang** [reply]
Hi Markus,

I tried to change `/map` to `/car/map`, but the problem was not solved.
![image](https://user-images.githubusercontent.com/39558023/121196315-aaa89780-c870-11eb-833a-a509af52e12d.png)

And I also tried other frameworks, and none of them reported errors.
![image](https://user-images.githubusercontent.com/39558023/121196793-112db580-c871-11eb-8e36-ed9486424c75.png)
![image](https://user-images.githubusercontent.com/39558023/121196890-21de2b80-c871-11eb-9a90-438cedcea317.png)
![image](https://user-images.githubusercontent.com/39558023/121197066-4934f880-c871-11eb-915f-568ef835bb33.png)

Should I try to add `map` to the current ts tree? If so, could you please tell me how to do it? 

Thanks!

**Reply by Lilly-yang** [reply]
It works now. I forgot to add `/car/map` in the sidebar. Thank you!
![image](https://user-images.githubusercontent.com/39558023/121230939-19e1b400-c890-11eb-9ae4-ec5873f8c584.png)

**Reply by Lilly-yang** [reply]
Hi,
I have successfully made a new map for my laboratory. Thank you!

I didn't go to the lab today, so I made another map for my current room, and then went back to the [Autonomous Navigation](https://mushr.io/tutorials/navigation/) tutorial. The good news is that the car and the real environment are synchronized with the new map on `rviz`. But I still cannot use the `2D Nav Goal` button to guide the car. No content was published in the `rhc` session, and the `"target set"` should have been published. In this session there is a warning `'Default map name being used, will be corrupted on map change. To fix, set '/map_file' parameter with map_file location'`. Here are all sessions (the last one is `rviz` session):
![image](https://user-images.githubusercontent.com/39558023/121389666-f3358300-c94c-11eb-9996-5e41ffa88cd4.png)
And here is the [recording](https://drive.google.com/file/d/17Px4PxuPl0wbww2lYteN2N4T8Vl67oZy/view?usp=sharing).

Could the warning be the reason why the navigation does not work?

Thanks in advance!

**Reply by schmittlema** [reply]
@Lilly-yang I think mushr_rhc is getting stuck waiting for the map. Could you try the following:
1. change [this argument](https://github.com/prl-mushr/mushr_rhc/blob/master/mushr_rhc_ros/launch/sim/sim.launch#L3) from 1 to 0
2. Change `/c2g/static_map` to `/map` [here](https://github.com/prl-mushr/mushr_rhc/blob/master/mushr_rhc_ros/launch/params/all_params.yaml)
3. before starting mushr_rhc run `roslaunch mushr_base map_server.launch map:=PATH TO YOUR MAP.yaml`

What this does is it replaces the map server mushr_rhc runs and makes manually sets the topics to align.

We are in the process of updating the autonomous navigation stack and should have a tutorial out soon that will alleviate much of these problems.

**Reply by Lilly-yang** [reply]
Thanks Matt. I tried but it still doesn't work. May I ask about when the new tutorial will be published?

**Reply by schmittlema** [reply]
Did the rhccontroller terminal output anything different than the image above? It should say "Initialized" if it is receiving a map. Is a map being published to `/map`? You can check via `rostopic hz /map`. The new tutorial will likely be ready within the next month potentially earlier

**Reply by Lilly-yang** [reply]
Hi Matt,
I am more aware of this function now. It's just that when I run the original  [mushr_rhc_ros map_server.launch](https://github.com/prl-mushr/mushr_rhc/blob/master/mushr_rhc_ros/launch/real/real.launch), it works and I can see the"`Initialized`". But when I change the  [map paths](https://github.com/prl-mushr/mushr_rhc/blob/077040568cb2d21fe03da863da137d0b7747174d/mushr_rhc_ros/launch/map_server.launch#L2)  to my own maps (two identical files), it doesn't work.
![image](https://user-images.githubusercontent.com/39558023/121928970-feb7ee00-cd40-11eb-8dc9-30ff0cf94ec2.png)
 The output of `rhc controller` terminal output as follows:
![image](https://user-images.githubusercontent.com/39558023/121927501-6a00c080-cd3f-11eb-9f55-3d66e4792c0a.png)
Could you please tell me what is the difference between `real-floor0-edited.yaml` and `real-floor0-edited-c2g-full-loop.yaml`? According to [pgm files](https://github.com/prl-mushr/mushr_rhc/tree/master/mushr_rhc_ros/maps) , I can't see any difference. 
Or do you have other suggestions? Thanks!

**Reply by schmittlema** [reply]
how big (image size + resolution in .yaml) is your inria map? It looks like the system is actually "working" and it initializing the psuedo planner which takes a while the first time it is done. Some of the intiali responses i posted [here](https://spectrum.chat/mushr/software/valid-points-len-xxx-runs-for-a-long-time~b4bba014-3adc-4b26-85e7-a816fcf74ce9) gets at this issue. TLDR crop all the unecessary unknown space in your map image and the controller should intialize faster. This is another thing we have removed in our coming controller update.

**Reply by Lilly-yang** [reply]
It works now. You are right, the system is running, it just I may need to be more patient. :o|

The information you posted [here](https://spectrum.chat/mushr/software/valid-points-len-xxx-runs-for-a-long-time~b4bba014-3adc-4b26-85e7-a816fcf74ce9) is very helpful. I adjusted the parameter from 3000 to 1500, but still, it took more than half an hour.

> What do you think is appropriate for different map size parameters?

My map is only the size of an ordinary room, about 20-30 square meters.
![image](https://user-images.githubusercontent.com/39558023/122060215-ab4fa980-cded-11eb-905c-662be7d08ed0.png)
I will try the 8-type map tomorrow. It is in a large room of about 100 square meters.

> Could you please tell me where I can set the speed?

The race car is running too fast, many times I don’t have time to connect the controller to stop it.

And sorry,

> what does TLDR mean?

Thanks a lot! :)

**Reply by schmittlema** [reply]
Yeah so your map image from your screenshot is 4000 x 4000 which is really really big (in terms of an image not in meters). So I would use [gimp](https://www.gimp.org/) or some other image editing tool to crop the unknown space in the image. Gmapping be default makes a really large unknown region when you build a map, which causes this problem. 

Make sure to connect the controller before starting the car. But you can set the speed [here](https://github.com/prl-mushr/mushr_rhc/blob/master/mushr_rhc_ros/launch/params/trajgen/dispersion.yaml#L10). 

[TLDR](https://www.dictionary.com/browse/tldr#:~:text=Tl%3Bdr%20or%20tldr%20stands,longer%20story%20or%20complicated%20phenomenon.) = Too Long Didn't Read. Its a reddit thing I think. You can think of it as equivalent to "in summary"

**Reply by abhisheksreesaila** [reply]
I have the exact same problem of my goal not being set  (pose is working), so i will try some of the suggestions above before posting here. 

Meanwhile I noticed the. "map" being big is a common problem.  I found "**crop_map**" tool that will reduce the resolution of the map without damaging the dimensions.  mine was reduced 200x200 (from 4000 x 4000)   You can checkout and download [here](https://ablearn.io/selfdriving/2020/09/07/FinalSteps.html#CROP-Map)

**Reply by abhisheksreesaila** [reply]
@schmittlema    I am having a similar issue. Screenshot attached.  RHC controller is not initialized.  I followed all the instructions above. Not sure. I could be missing a very minor config 

------------------------------------------

![pre-req --- File Exists in RHC Map Server](https://user-images.githubusercontent.com/11252442/128646974-de716e31-bd38-4c41-856b-4189a9950477.png)

![pre-req --- MUSHR RHC ROS Launch](https://user-images.githubusercontent.com/11252442/128646975-6a30d03b-7327-444c-af1d-04c59ee3aeba.png)

------------------------------------------
![3  Goal is set in RVIZ](https://user-images.githubusercontent.com/11252442/128646976-9506c6b6-8676-4753-b297-6970e64e2be0.png)
![2  RHC topics](https://user-images.githubusercontent.com/11252442/128646978-bf413a5d-c7e4-4e8e-8a2e-490aeee34dbd.png)
![0  POSE set is success](https://user-images.githubusercontent.com/11252442/128646979-00f47dec-6f92-47c8-bd12-ea724989efc2.png)
![1  Warning Issue](https://user-images.githubusercontent.com/11252442/128646980-a74c707a-2cf2-4f31-84fa-2ed1e1c6f6d8.png)
)

**Reply by abhisheksreesaila** [reply]
**I guess the warning issue.  I just started the RHC map server.** 

--------------------------
--------------------------

![Warnings fixed](https://user-images.githubusercontent.com/11252442/128647429-bc790944-5f93-4fa3-8d39-1448de304439.png)
![Checked Map_file Param](https://user-images.githubusercontent.com/11252442/128647430-5638e94d-f397-4435-94d8-21e146e49c3c.png)
![Started RHC Mapserver](https://user-images.githubusercontent.com/11252442/128647431-2a03e5e1-3748-4e8f-bce8-0cfadfbe5fc1.png)

**Reply by abhisheksreesaila** [reply]
@schmittlema  Now all left is the issue with goal not being set.   i guess the messages are not coming over. **_POSE is set, but GOAL is not.  Weird.  Any ideas?_** 

--------------------------------

![Screen Shot 2021-08-08 at 3 26 04 PM](https://user-images.githubusercontent.com/11252442/128647470-3c94b2b3-529f-4da9-9cfe-93e2d565df1e.png)

**Reply by schmittlema** [reply]
@abhisheksreesaila did the rhc_controller report it was initialized? Presumably after some messages "Valid len #". If so, are you running this on the real platform? And if you are, is the ROS_IP properly set? ROS_IP environment variable on the car should be the car's IP address, and on the computer with rviz it should be the computer's IP address

**Reply by abhisheksreesaila** [reply]
- Yes, I am running on the real car
-  RHC controller did NOT initialize. 
    -  I don't get the "Valid Len # ...."
- Yes the IP's are set correctly. I know that because the POSE messages are published correctly (from the laptop to the car)
   - I see the print statements on the particle filter component as "Setting Pose"
![Screen Shot 2021-08-08 at 12 01 12 PM](https://user-images.githubusercontent.com/11252442/128908884-9e3d65d1-5c6a-4ba4-b397-bc7b0bbc9689.png)

**Reply by abhisheksreesaila** [reply]
I will say that - whenever i set the pose the car,  the car on the rviz slides to the edge of the map.  let me post a video to show that..

**Reply by schmittlema** [reply]
if the rhc controller did not initialize then it is probably stuck waiting on the map. could you paste the result of `rosservice list`?

**Reply by abhisheksreesaila** [reply]
That what I thought from your comments above.  I thought the map file param was not set, then if fixed the launch to start the map server on the RHC node as well. 


Here is the list results of    `rosservice list` 

![part1](https://user-images.githubusercontent.com/11252442/128931762-0bed2106-f549-4e9d-bf0f-7cdaab445816.png)
![part2](https://user-images.githubusercontent.com/11252442/128931767-21d73d6d-bf20-4456-8c97-b4c907149392.png)

**Reply by abhisheksreesaila** [reply]
RHC NODE

![Screen Shot 2021-08-10 at 1 37 39 PM](https://user-images.githubusercontent.com/11252442/128932028-189d179e-7cdc-414e-9616-db609b113b53.png)

**Reply by schmittlema** [reply]
ok it looks like the service is `/static_map` but your param is set to `/map`. Could you try chaning to `/static_map`. That param should be set in `all_params.yaml`

**Reply by abhisheksreesaila** [reply]
Yes that worked,  it initializes and I am able to set the goal.  
![Screen Shot 2021-08-10 at 8 11 14 PM](https://user-images.githubusercontent.com/11252442/128978481-99f544b4-3cf2-4afe-91b8-84e881fe28d9.png)

**Reply by schmittlema** [reply]
awesome, is there anything else causing trouble? I saw you mentioned something about the car sliding in rviz? That would likely be related to the particle filter

**Reply by abhisheksreesaila** [reply]
Thanks @schmittlema   It is sliding. See video attached.  but i am able to run the car autonomously. 

https://user-images.githubusercontent.com/11252442/129122373-7291ce67-f823-4fdc-bba3-520724038a6f.mov

![Screen Shot 2021-08-11 at 5 38 16 PM](https://user-images.githubusercontent.com/11252442/129122391-11b269ec-ef83-46fe-8715-bc15d0253d93.png)

**Reply by abhisheksreesaila** [reply]
@schmittlema   the only major problem i am trying to solve before i take this to street is the battery problem.  nvidia jetson switches off after 1-2 min of charging battery.  the converter shows 5V, but jetson struggles to keep up. have you guys had any issues with it.    VESC and its battery seems fine.

**Reply by schmittlema** [reply]
Sliding problem:
 - The particle filter is trying to localize the car in the map using the laser scan (pink dots) which are coming from the real scanner. It looks like from the scan that your car is sitting in some corner. So when you try to localize it to another part of the map it will move untill it finds a spot that looks like the corner the real car is sitting in. Could you try setting the 2D pose estimate near the corner the car is truely in? I would also try to roughly match orientation.
 
 Vesc problem:
- Let's open up a new discussionn for this as it does not pertain to navigation. We can certainly help you from there.

**Reply by abhisheksreesaila** [reply]
Understood.  Yes the car is sitting the one corner connected to the power source since the battery is not working as expected.
